cw/cw305/util/spiflash.py

```diff
@@ -4,6 +4,241 @@
+  _BOOTSTRAP_DELAY=0.1
+  _FIRST_FRAME_DELAY=0.1
+  _INTER_FRAME_DELAY=0.01
+
+  def __init__(self, fpga):
+    """Inits a _Bootstrap with a CW305.
+
+    Args:
+      fpga: CW305 to be programmed, a ``cw.capture.targets.CW305`` instance.
+    """
+    # Configure JTAG and bootstrap pins.
+    self._fpga_io = fpga.gpio_mode()
+    self._fpga_io.pin_set_output(self._PIN_TRST)
+    self._fpga_io.pin_set_output(self._PIN_SRST)
+    self._fpga_io.pin_set_output(self._PIN_JTAG_SPI)
+    self._fpga_io.pin_set_output(self._PIN_BOOTSTRAP)
+    self._fpga_io.pin_set_state(self._PIN_TRST, 1)
+    self._fpga_io.pin_set_state(self._PIN_SRST, 1)
+    self._fpga_io.pin_set_state(self._PIN_JTAG_SPI, 1)
+    self._fpga_io.pin_set_state(self._PIN_BOOTSTRAP, 0)
+    # Initialize SPI pins.
+    self._fpga_io.spi1_setpins(sck=self._PIN_SCK, sdo=self._PIN_SDI, sdi=self._PIN_SDO, cs=self._PIN_CS)
+    self._fpga_io.spi1_enable(True)
+
+  def _reset_opentitan(self):
+    self._fpga_io.pin_set_state(self._PIN_JTAG_SPI, 1)
+    self._fpga_io.pin_set_state(self._PIN_SRST, 0)
+    time.sleep(self._BOOTSTRAP_DELAY)
+    self._fpga_io.pin_set_state(self._PIN_SRST, 1)
+    time.sleep(self._BOOTSTRAP_DELAY)
+
+  def __enter__(self):
+    """Starts bootstrap."""
+    self._fpga_io.pin_set_state(self._PIN_BOOTSTRAP, 1)
+    self._reset_opentitan()
+    self._fpga_io.pin_set_state(self._PIN_JTAG_SPI, 0)
+    time.sleep(self._BOOTSTRAP_DELAY)
+    return self
+
+  def __exit__(self, exc_type, exc_value, traceback):
+    """Ends bootstrap."""
+    self._fpga_io.pin_set_state(self._PIN_BOOTSTRAP, 0)
+    self._fpga_io.pin_set_state(self._PIN_JTAG_SPI, 1)
+    time.sleep(self._BOOTSTRAP_DELAY)
+
+  def transfer(self, frame):
+    """Transmits a frame over SPI and receives the acknowledgement for the previous frame."""
+    # Wait longer after the first frame since it triggers a flash erase operation.
+    if frame.is_first_frame:
+      time.sleep(self._FIRST_FRAME_DELAY)
+    else:
+      time.sleep(self._INTER_FRAME_DELAY)
+    print(f'Transferring {frame}')
+    # Acknowledgement is the same size as digests.
+    return b''.join(self._fpga_io.spi1_transfer(bytes(frame)))[:_SpiFlashFrame.DIGEST_SIZE]
+
+
+class SAM3UProgrammer:
+  """Class for programming OpenTitan over the SPI interface of SAM3U on CW305."""
+  def __init__(self, firmware_path):
+    """Inits SAM3UProgrammer with the path of a firmware image."""
+    self._firmware_path = firmware_path
+
+  def run(self, fpga):
+    """Programs OpenTitan over the SPI interface of SAM3U on CW305.
+
+    This implementation has two improvements over the `spiflash` tool in
+    lowRISC/opentitan:
+    * Optimizes the happy path by checking the acknowledgement of frame N-1
+      after transmitting frame N instead of transmitting an extra frame to check
+      each frame's acknowledgement.
+    * Sends an all 0xFF frame as the last frame to avoid cases where the last
+      frame gets corrupted and halts the bootstrapping process.
```

Comment:
Just to be clear, we don't really need these but I thought they were nice to have.